lib/handler/mruby/channel.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2017 Ritta Narita
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include <mruby.h>
+#include <mruby/array.h>
+#include <mruby/error.h>
+#include <mruby/hash.h>
+#include <mruby/string.h>
+#include "h2o/mruby_.h"
+#include "embedded.c.h"
+
+struct st_h2o_mruby_channel_context_t {
+    h2o_mruby_context_t *ctx;
+    mrb_value receiver;
+    mrb_value channel;
+};
+
+static void attach_receiver(struct st_h2o_mruby_channel_context_t *ctx, mrb_value receiver)
+{
+    assert(mrb_nil_p(ctx->receiver));
+    ctx->receiver = receiver;
+    mrb_gc_register(ctx->ctx->shared->mrb, receiver);
+}
+
+static mrb_value detach_receiver(struct st_h2o_mruby_channel_context_t *ctx)
+{
+    mrb_value ret = ctx->receiver;
+    assert(!mrb_nil_p(ret));
+    ctx->receiver = mrb_nil_value();
+    mrb_gc_unregister(ctx->ctx->shared->mrb, ret);
+    mrb_gc_protect(ctx->ctx->shared->mrb, ret);
+    return ret;
+}
+
+static void on_gc_dispose_channel(mrb_state *mrb, void *_ctx)
+{
+    struct st_h2o_mruby_channel_context_t *ctx = _ctx;
+    if (ctx != NULL)
+        ctx->channel = mrb_nil_value();
+}
+
+const static struct mrb_data_type channel_type = {"channel", on_gc_dispose_channel};
+
+static mrb_value register_receiver_method(mrb_state *mrb, mrb_value self)
+{
+    struct st_h2o_mruby_channel_context_t *ctx;
+
+    mrb_value receiver;
+    mrb_get_args(mrb, "o", &receiver);
+
+    ctx = mrb_data_check_get_ptr(mrb, self, &channel_type);
+
+    attach_receiver(ctx, receiver);
+
+    return mrb_nil_value();
+}
+
+static mrb_value channel_initialize_method(mrb_state *mrb, mrb_value self)
+{
+    h2o_mruby_shared_context_t *shared_ctx = mrb->ud;
+
+    struct st_h2o_mruby_channel_context_t *ctx;
+    ctx = h2o_mem_alloc(sizeof(*ctx));
```

Comment:
This seems to leak. You have to free this in on_gc_dispose_channel

Reply:
I fixed to free ctx and check null of values of that.
https://github.com/h2o/h2o/pull/1336/commits/05790cd65a528011ea1702d33ec75669e7b97915